src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs

```diff
@@ -245,17 +246,36 @@ private static void RecordRequestStartEventLog(HttpContext httpContext)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private Activity StartActivity(HttpContext httpContext, out bool hasDiagnosticListener)
+        private Activity? StartActivity(HttpContext httpContext, bool loggingEnabled, bool diagnosticListenerActivityCreationEnabled, out bool hasActivityStartDiagnosticListener)
         {
-            var activity = new Activity(ActivityName);
-            hasDiagnosticListener = false;
+            hasActivityStartDiagnosticListener = false;
+            if (diagnosticListenerActivityCreationEnabled)
+            {
+                if (_diagnosticListener.IsEnabled(ActivityStartKey))
+                {
+                    hasActivityStartDiagnosticListener = true;
+                }
+            }
+            // Short-circuit to avoid doing an expensive header lookup
+            if (!(diagnosticListenerActivityCreationEnabled || _activitySource.HasListeners() || loggingEnabled))
```

Comment:
would it make sense to move this check in the beginning of the method?

Reply:
Yeah. About to push a change 😄